Source/ExecutionEngine/ExecutionEngine.cs

```diff
@@ -571,16 +571,11 @@ public void Inline(Program program)
       }
 
       var outcome = await VerifyEachImplementation(output, processedProgram, stats, programId, er, requestId, stablePrioritizedImpls);
-      if (Options.PrintPassive) {
-        Options.PrintUnstructured = 1;
-        PrintBplFile(Options.PrintFile, processedProgram.Program, true, true, Options.PrettyPrint);
-      }
 
-      if (Options.LeanFile is not null) {
-        var writer = new StreamWriter(Options.LeanFile);
-        LeanGenerator.EmitPassiveProgramAsLean(Options, processedProgram.Program, writer);
-        writer.Close();
-      }
+      // We unfortunately have to do any processing of the passive program after verification,
```

Comment:
I don't understand this comment. These lines seems contradictory, both saying that `processedProgram.Program` is an already passified program, and that it's not yet passified because that only happens during verification.

By the way, what step do you mean by "verification" in the context of Boogie? The whole pipeline could be considered verification.

Reply:
What I mean is that `ProcessPassiveProgram` needs to be called after `VerifyEachImplementation` instead of before, because `VerifyEachImplementation` turns `processedProgram` into a passive program, destructively. I would love it if everything happened in `PreProcessProgramVerification`, but it doesn't right now, and I think it'd take a pretty big refactoring to get it there.

Reply:
Ah, if you change `verification` to `calling VerifyEachImplementation` in the comment, then it would be more clear to me.

Reply:
Btw, extracting the passive-ication is sort of already done in the method `vcGenerator.PrepareImplementation`, only it only does it for a single implementation

Reply:
It would generally make sense to generate Lean code as an alternative to SMT-Lib, so it could make sense for it to be called from the `VCGeneration` package. I took a stab at refactoring things to do that, but as things currently stand there are two issues:

* The Lean generator depends on the `VCGeneration` package, and we can't create a cyclic dependency.
* The Lean generator currently translates the whole program in one go, which doesn't really fit with the current structure of the VC generator.

I think it's possible to refactor things further to not require that cyclic dependency, and I'm considering whether the Lean generator should work on a per-implementation basis, but I think I'd prefer to leave both for a future PR (probably one in the pretty near future, though).

Reply:
To me it's already a problem that `ExecutionEngine` is calling out to `LeanGenerator`, because it bloats the scope of `ExecutionEngine`.

I would go for the combination of my previous two suggestions, to remove the dependency of ExecutionEngine on LeanGenerator, and make use of `vcGenerator.PrepareImplementation`

Change

```c#
var processedProgram = PreProcessProgramVerification(program);
if (!Options.Verify)
{
  return PipelineOutcome.Done;
}

if (Options.ContractInfer)
{
  return await RunHoudini(program, stats, er);
}

...
var outcome = await VerifyEachImplementation(output, processedProgram, stats, programId, er, requestId, stablePrioritizedImpls);

if (Options.LeanFile is not null) {
  var writer = new StreamWriter(Options.LeanFile);
  LeanAutoGenerator.EmitPassiveProgramAsLean(Options, passiveProgram, writer);
  writer.Close();
}

```

Into
```c#
var processedProgram = PreProcessProgramVerification(program);
if (!Options.Verify)
{
  return PipelineOutcome.Done;
}

if (Options.UseResolvedProgram != null && Options.UseResolvedProgram(processedProgram))
{
  return;
}

...
var outcome = await VerifyEachImplementation(output, processedProgram, stats, programId, er, requestId, stablePrioritizedImpls);
```

Note that both Houdini and Lean are no longer mentioned.

Then instead of setting `LeanFile`, set `Options.UseResolvedProgram` with:

```c#
void LetLeanUseUseResolvedProgram(Program program) {
  foreach(var implementation in program.Implementations) {
    vcGenerator.PrepareImplementation(new ImplementationRun(implementation, options.OutputWriter), ..)
  }

  var writer = new StreamWriter(Options.LeanFile);
  EmitPassiveProgramAsLean(Options, program, writer);
  writer.Close();
}

```

Reply:
I like this approach, and I've pushed some changes that go in roughly that direction.